pkgs/build-support/docker/default.nix

```diff
@@ -886,4 +888,70 @@ rec {
         done;
       '' + extraCommands;
     });
+
+  # This function builds a docker image that behaves like a nix-shell
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name
+    , drv
+    , tag ? null
+    # Extra packages available inside the image
+    , extraPackages ? []
+    # Extra bash commands to execute when the image is ran
+    , extraInit ? ""
+    }:
+    # Remove once https://github.com/NixOS/nixpkgs/pull/77952 is merged
+    let
+
+      finalDrv = drv.overrideAttrs (old: {
+        nativeBuildInputs = old.nativeBuildInputs or [] ++ extraPackages;
+      });
+
+      setupPackage = runCommandNoCC "setup" {
+        profile = ''
+          # This mimics part of nix-shell's initialization
+          source $stdenv/setup
+          set +e
+          eval "$shellHook"
+          cd /home
+        '' + extraInit;
+        passAsFile = [ "profile" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/{build,outputs}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
+        ln -s ${coreutils}/bin/env $out/usr/bin/env
+        cp "$profilePath" $out/etc/profile
+        ln -s "${cacert}/etc/ssl/certs/ca-bundle.crt" $out/etc/ssl/certs/ca-certificates.crt
+      '';
+    in buildLayeredImage {
+      inherit name tag;
+      contents = [ setupPackage ];
+      # These folders should be writable
+      extraCommands = "chmod +w -R tmp home";
+
+      # Don't run this image as root because many programs don't like that
+      config.User = "1000:1000";
+      # Set the owner of the files to that user as well
+      uid = 1000;
+      gid = 1000;
+
+      # Pass --login such that /etc/profile gets loaded
+      # Use bashInteractive for tab completion support and such
+      config.Entrypoint = [ "${bashInteractive}/bin/bash" "--login" ];
+      config.Env = let
+        # We can't just use `toString` on all derivation attributes because that
+        # would not put path literals in the closure. So we explicitly copy
+        # those into the store here
+        stringValue = value: if builtins.typeOf value == "path" then "${value}"
+          else if builtins.isList value then toString (map stringValue value)
+          else toString value;
+      in lib.mapAttrsToList (name: value: "${name}=${stringValue value}") (finalDrv.drvAttrs // {
+        # These are needed for stdenv setup to work
+        NIX_STORE = "/nix/store";
+        NIX_BUILD_TOP = "/home/build";
+
+        HOME = "/home";
```

Comment:
I'd prefer `/home/user`, to match FHS expectations.
Also `/home/build` -> `/home/user/build`, etc.